prov/rxd/src/rxd_proto.h

```diff
@@ -88,36 +88,59 @@ struct rxd_cts_pkt {
 
 struct rxd_ack_pkt {
 	struct rxd_base_hdr	base_hdr;
-	struct rxd_pkt_hdr	pkt_hdr;
+	struct rxd_ext_hdr	ext_hdr;
 	//TODO fill in more fields? Selective ack?
 };
 
-//TODO split this into separate packet types to not waste space
-struct rxd_op_pkt {
+struct rxd_inline_pkt {
 	struct rxd_base_hdr	base_hdr;
-	struct rxd_pkt_hdr	pkt_hdr;
-
-	uint64_t		num_segs;
-	union {
-		uint64_t		tag;
-		struct {
-			uint64_t		iov_count;
-			struct ofi_rma_iov	rma[RXD_IOV_LIMIT];
-
-			uint32_t		datatype;
-			uint32_t		atomic_op;
-		};
-	};
+	char			msg[];
+};
 
+struct rxd_op_hdr {
 	uint64_t		cq_data;
 	uint64_t		size;
+	uint32_t		num_segs;
+	uint16_t		flags;
+	uint16_t		tx_id;
+};
+
+struct rxd_msg_pkt {
+	struct rxd_base_hdr	base_hdr;
+	struct rxd_op_hdr	op_hdr;
+
+	uint64_t		tag;
 
 	char			msg[];
-}; 
+};
+
+struct rxd_rma_pkt {
+	struct rxd_base_hdr	base_hdr;
+	struct rxd_op_hdr	op_hdr;
+
+	uint64_t		iov_count;
+
+	struct ofi_rma_iov	rma[RXD_IOV_LIMIT];
```

Comment:
Since you carry the iov_count, you only need to carry the needed number of iovecs, rather than always carrying the full iov array.